clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -1134,18 +1138,21 @@ public void handle(OffsetCommitResponse commitResponse, RequestFuture<Void> futu
                             future.raise(error);
                             return;
                         } else if (error == Errors.REBALANCE_IN_PROGRESS) {
-                            /* Consumer never tries to commit offset in between join-group and sync-group,
+                            /* Consumer should not try to commit offset in between join-group and sync-group,
                              * and hence on broker-side it is not expected to see a commit offset request
                              * during CompletingRebalance phase; if it ever happens then broker would return
-                             * this error. In this case we should just treat as a fatal CommitFailed exception.
-                             * However, we do not need to reset generations and just request re-join, such that
-                             * if the caller decides to proceed and poll, it would still try to proceed and re-join normally.
+                             * this error to indicate that we are still in the middle of a rebalance.
+                             * In this case we would throw a RetriableCommitFailedException,
+                             * request re-join but do not reset generations. If the callers decide to retry they
+                             * can go ahead and call poll to finish up the rebalance first, and then try commit again.
                              */
                             requestRejoin();
-                            future.raise(new CommitFailedException());
+                            future.raise(new RetriableCommitFailedException("Offset commit cannot be completed since the " +
+                                "consumer group is executing a rebalance at the moment. You can try completing the rebalance " +
+                                "by calling poll() and then retry commit again"));
                             return;
-                        } else if (error == Errors.UNKNOWN_MEMBER_ID
-                                || error == Errors.ILLEGAL_GENERATION) {
+                        } else if (error == Errors.UNKNOWN_MEMBER_ID ||
+                                   error == Errors.ILLEGAL_GENERATION) {
```

Comment:
Just want to check my understanding. The user may attempt to commit offsets while the broker has a JoinGroup in purgatory. In this case, we would send the older generation which would be doomed to fail with `ILLEGAL_GENERATION` once the join completes. In this case, should we still reset the generation as we do below? I am wondering if it is useful to remember the generation that an offset commit was sent with (perhaps inside `OffsetCommitCompletion`) so that we only reset if necessary.

Reply:
Yes we are doomed to fail in this case, but I think it is no harm to reset generation in this case still since all it does is to clear the generation id and set state to `UNJOINED` which should be the case anyways since we've sent the join-group request (we set the state to `UNJOINED` before sending it).

Reply:
But it would result in an unnecessary JoinGroup retry after we reset the generation, right?

Reply:
My understanding is that we only `rejoinNeeded = false;` in the outer caller, which means we would only reset it to false after the sync-group is received. So before that, no matter how many times we change it to `true` it would not matter?

Reply:
EDIT: Yes you're right, I think it would cause the sync-group's callback to still `failRebalanceAndRejoin`. Would change this part accordingly.